listenbrainz/webserver/static/js/src/brainzplayer/ProgressBar.tsx

```diff
@@ -8,17 +10,26 @@ type ProgressBarProps = {
 };
 
 // How many milliseconds to navigate to with keyboard left/right arrows
-const keyboardStepMS: number = 5000;
-const keyboardBigStepMS: number = 10000;
+const KEYBOARD_BIG_STEP_MS: number = 5000;
+const KEYBOARD_STEP_MS: number = 10000;
+
+const EVENT_KEY_ARROWLEFT: string = "ArrowLeft";
+const EVENT_KEY_ARROWRIGHT: string = "ArrowRight";
+const EVENT_TYPE_CLICK: string = "click";
+const EVENT_TYPE_MOUSEMOVE: string = "mousemove";
+
+const TOOLTIP_INITIAL_CONTENT: string = "0:00";
+const TOOLTIP_TOP_OFFSET: number = 102;
 
 const ProgressBar = (props: ProgressBarProps) => {
   const { durationMs, progressMs, seekToPositionMs } = props;
+  const [tipContent, setTipContent] = React.useState(TOOLTIP_INITIAL_CONTENT);
   const progressPercentage = Math.round(
     Number((progressMs * 100) / durationMs)
   );
   const hideProgressBar = isNaN(progressPercentage) || progressPercentage <= 0;
 
-  const progressClickHandler = (
+  const mouseEventHandler = (
```

Comment:
Wow that sounds interesting! The thought of "firing million times a second" did dawn on me but I didn't know it can be throttled. I'd love some more explanation please :)

Reply:
Great :)
OK, React hats on everybody!

So the [throttle](https://lodash.com/docs/4.17.15#throttle) function from lodash takes a function as first argument and a number of milliseconds as second argument.
It returns another function which no matter how many times you call it will only be fired once every X milliseconds. It can be canceled too.

The second part is that we don't want to recreate that function on every render (the delay functionality wouldn't work at all…), so in order to memoize it we use the [`useCallback` React hook](https://reactjs.org/docs/hooks-reference.html#usecallback) so that the throttled function is only recreated when `durationMs`,`seekToPositionMs` or `setTipContent` change (the rule of thumb is: if you use a prop or setState in a useCallback or useEffect hook, add it to the array of dependencies [the second argument to the hook]).

Altogether, it would look something like this:

```tsx
const mouseEventHandler = React.useCallback(( event: React.MouseEvent<HTMLInputElement>)=>{
    // Creating our actual handler function inside useCallback for memoization
    const handler = function () {
        const progressBarBoundingRect = event.currentTarget.getBoundingClientRect();
        const progressBarWidth = progressBarBoundingRect.width;
        const musicPlayerXOffset = progressBarBoundingRect.x;
        const absoluteClickXPos = event.clientX;
        const relativeClickXPos = absoluteClickXPos - musicPlayerXOffset;
        const percentPos = relativeClickXPos / progressBarWidth;
        const positionMs = Math.round(durationMs * percentPos);
        const positionTime = msToTime(positionMs);
    
        if (event.type === EVENT_TYPE_MOUSEMOVE) {
          setTipContent(positionTime);
        } else if (event.type === EVENT_TYPE_CLICK) {
          seekToPositionMs(positionMs);
        }
      };
      // lodash's throttle returns a throttled function
      const throttledHandler = throttle(handler,350);
      // Invoke the throttled function, will only fire once every X millisecond
      throttledHandler();
      // For hook cleanup, return the throttled function's cancel method
      return throttledHandler.cancel
  },[durationMs,seekToPositionMs,setTipContent])
```


Note: I didn't test his code, so there's no guarantee it will work, but it should put you in the right direction.
For testing how often what is called, you can put console logs (or better yet logpoints in the browser console): one outside `handler` and one inside.
The one outside should be printed very often while the one inside once every 350 milliseconds.

You might also want to play with that value; I think under 350ms humans don't really notice differences, but we might be able to increase the delay a bit for better performance.
We just need it to be fluid enough that user's won't think "huh, why is this slow?"

Reply:
Thank you so much for this explanation! I ended up diving deep into debounce and throttling in JS, and reading more about `useCallBack`.
However, I couldn't get throttle to work as we'd expected. When I finally thought it worked as expected, it actually ended up behaving the same as the original function 😅 
So I decided to use a custom `useThrottle()` hook which is originally from here - https://stackoverflow.com/a/62017005. I got it working AND the code looks clean :)

Reply:
I love a good deep dive myself, you get to learn useful tools that could help solve another problem down the line :)

Reply:
Definitely 💯